File: src/services/settingsService.ts
import { api } from './api';
import { supabase, isSupabaseConfigured } from '../lib/supabaseClient';

export interface EventConfig {
  eventName: string;
  eventDate: string;
  targetDateIso: string;
  registrationStatus: 'open' | 'closed' | 'paused';
  registrationOpenDate: string;
  registrationCloseDate: string;
  websiteVisibility: 'published' | 'maintenance';
  eventStatus: 'planning' | 'active' | 'completed';
}

export interface NotificationConfig {
  announcementEmailsEnabled: boolean;
  registrationConfirmationEmailsEnabled: boolean;
  eventReminderAlertsEnabled: boolean;
  emailSenderName: string;
  providerStatus: 'ready' | 'simulated' | 'configured';
}

export interface SystemConfig {
  maintenanceMode: boolean;
  maxRegistrations: number;
  announcementBannerEnabled: boolean;
  announcementBannerText: string;
  debugLogging: boolean;
}

export interface AdminProfile {
  displayName: string;
  email: string;
  role: string;
  accountStatus: string;
  lastSignIn?: string | null;
  authProvider: string;
}

export interface FullSettings {
  event: EventConfig;
  notifications: NotificationConfig;
  system: SystemConfig;
  adminProfile: AdminProfile;
}

export interface UserRoleItem {
  id: string;
  userEmail: string;
  role: 'admin' | 'coordinator' | 'jury' | 'participant';
  displayName: string;
  department: string;
  isActive: boolean;
  assignedBy: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface AuditLogItem {
  id: string;
  action: string;
  performedBy: string;
  target: string;
  details: string;
  createdAt?: string;
}

export const DEFAULT_FULL_SETTINGS: FullSettings = {
  event: {
    eventName: 'PRAGATHI 2K26',
    eventDate: '09 October 2026',
    targetDateIso: '2026-10-09T09:00:00+05:30',
    registrationStatus: 'open',
    registrationOpenDate: '2026-08-01T00:00:00+05:30',
    registrationCloseDate: '2026-10-01T23:59:59+05:30',
    websiteVisibility: 'published',
    eventStatus: 'active',
  },
  notifications: {
    announcementEmailsEnabled: true,
    registrationConfirmationEmailsEnabled: true,
    eventReminderAlertsEnabled: true,
    emailSenderName: 'PRAGATHI 2K26 Secretariat',
    providerStatus: 'ready',
  },
  system: {
    maintenanceMode: false,
    maxRegistrations: 500,
    announcementBannerEnabled: false,
    announcementBannerText: 'Welcome to PRAGATHI 2K26 Expo Registration Portal!',
    debugLogging: false,
  },
  adminProfile: {
    displayName: 'Lead Administrator',
    email: 'admin@sru.edu.in',
    role: 'admin',
    accountStatus: 'verified',
    lastSignIn: null,
    authProvider: 'email',
  },
};

export const DEFAULT_USER_ROLES: UserRoleItem[] = [
  {
    id: 'role-admin-1',
    userEmail: 'admin@sru.edu.in',
    role: 'admin',
    displayName: 'Lead Administrator',
    department: 'Deanery',
    isActive: true,
    assignedBy: 'system',
    createdAt: '2026-08-01T00:00:00Z',
  },
  {
    id: 'role-coord-1',
    userEmail: 'coordinator@sru.edu.in',
    role: 'coordinator',
    displayName: 'Chief Event Coordinator',
    department: 'CSE',
    isActive: true,
    assignedBy: 'system',
    createdAt: '2026-08-01T00:00:00Z',
  },
  {
    id: 'role-jury-1',
    userEmail: 'jury.head@sru.edu.in',
    role: 'jury',
    displayName: 'Evaluation Panel Head',
    department: 'Research',
    isActive: true,
    assignedBy: 'system',
    createdAt: '2026-08-01T00:00:00Z',
  },
];

function transformFromBackend(data: any): FullSettings {
  if (!data) return DEFAULT_FULL_SETTINGS;
  const ev = data.event || {};
  const notif = data.notifications || {};
  const sys = data.system || {};
  const adm = data.admin_profile || data.adminProfile || {};

  return {
    event: {
      eventName: ev.event_name ?? DEFAULT_FULL_SETTINGS.event.eventName,
      eventDate: ev.event_date ?? DEFAULT_FULL_SETTINGS.event.eventDate,
      targetDateIso: ev.target_date_iso ?? DEFAULT_FULL_SETTINGS.event.targetDateIso,
      registrationStatus: ev.registration_status ?? DEFAULT_FULL_SETTINGS.event.registrationStatus,
      registrationOpenDate: ev.registration_open_date ?? DEFAULT_FULL_SETTINGS.event.registrationOpenDate,
      registrationCloseDate: ev.registration_close_date ?? DEFAULT_FULL_SETTINGS.event.registrationCloseDate,
      websiteVisibility: ev.website_visibility ?? DEFAULT_FULL_SETTINGS.event.websiteVisibility,
      eventStatus: ev.event_status ?? DEFAULT_FULL_SETTINGS.event.eventStatus,
    },
    notifications: {
      announcementEmailsEnabled: notif.announcement_emails_enabled ?? DEFAULT_FULL_SETTINGS.notifications.announcementEmailsEnabled,
      registrationConfirmationEmailsEnabled: notif.registration_confirmation_emails_enabled ?? DEFAULT_FULL_SETTINGS.notifications.registrationConfirmationEmailsEnabled,
      eventReminderAlertsEnabled: notif.event_reminder_alerts_enabled ?? DEFAULT_FULL_SETTINGS.notifications.eventReminderAlertsEnabled,
      emailSenderName: notif.email_sender_name ?? DEFAULT_FULL_SETTINGS.notifications.emailSenderName,
      providerStatus: notif.provider_status ?? DEFAULT_FULL_SETTINGS.notifications.providerStatus,
    },
    system: {
      maintenanceMode: sys.maintenance_mode ?? DEFAULT_FULL_SETTINGS.system.maintenanceMode,
      maxRegistrations: sys.max_registrations ?? DEFAULT_FULL_SETTINGS.system.maxRegistrations,
      announcementBannerEnabled: sys.announcement_banner_enabled ?? DEFAULT_FULL_SETTINGS.system.announcementBannerEnabled,
      announcementBannerText: sys.announcement_banner_text ?? DEFAULT_FULL_SETTINGS.system.announcementBannerText,
      debugLogging: sys.debug_logging ?? DEFAULT_FULL_SETTINGS.system.debugLogging,
    },
    adminProfile: {
      displayName: adm.display_name ?? DEFAULT_FULL_SETTINGS.adminProfile.displayName,
      email: adm.email ?? DEFAULT_FULL_SETTINGS.adminProfile.email,
      role: adm.role ?? DEFAULT_FULL_SETTINGS.adminProfile.role,
      accountStatus: adm.account_status ?? DEFAULT_FULL_SETTINGS.adminProfile.accountStatus,
      lastSignIn: adm.last_sign_in ?? DEFAULT_FULL_SETTINGS.adminProfile.lastSignIn,
      authProvider: adm.auth_provider ?? DEFAULT_FULL_SETTINGS.adminProfile.authProvider,
    },
  };
}

function transformToBackend(settings: Partial<FullSettings>): Record<string, any> {
  const payload: Record<string, any> = {};
  if (settings.event) {
    payload.event = {
      event_name: settings.event.eventName,
      event_date: settings.event.eventDate,
      target_date_iso: settings.event.targetDateIso,
      registration_status: settings.event.registrationStatus,
      registration_open_date: settings.event.registrationOpenDate,
      registration_close_date: settings.event.registrationCloseDate,
      website_visibility: settings.event.websiteVisibility,
      event_status: settings.event.eventStatus,
    };
  }
  if (settings.notifications) {
    payload.notifications = {
      announcement_emails_enabled: settings.notifications.announcementEmailsEnabled,
      registration_confirmation_emails_enabled: settings.notifications.registrationConfirmationEmailsEnabled,
      event_reminder_alerts_enabled: settings.notifications.eventReminderAlertsEnabled,
      email_sender_name: settings.notifications.emailSenderName,
      provider_status: settings.notifications.providerStatus,
    };
  }
  if (settings.system) {
    payload.system = {
      maintenance_mode: settings.system.maintenanceMode,
      max_registrations: settings.system.maxRegistrations,
      announcement_banner_enabled: settings.system.announcementBannerEnabled,
      announcement_banner_text: settings.system.announcementBannerText,
      debug_logging: settings.system.debugLogging,
    };
  }
  if (settings.adminProfile) {
    payload.admin_profile = {
      display_name: settings.adminProfile.displayName,
      email: settings.adminProfile.email,
      role: settings.adminProfile.role,
      account_status: settings.adminProfile.accountStatus,
    };
  }
  return payload;
}

export async function getAdminSettings(): Promise<FullSettings> {
  try {
    const res = await api.admin.getSettings();
    if (res && res.data) {
      return transformFromBackend(res.data);
    }
  } catch (err) {
    console.warn('[settingsService] getAdminSettings FastAPI fallback:', err);
  }

  if (isSupabaseConfigured && supabase) {
    try {
      const { data, error } = await supabase.from('system_settings').select('*');
      if (!error && data && data.length > 0) {
        const merged: Record<string, any> = {};
        for (const row of data) {
          if (row.key === 'event_config') merged.event = row.value;
          if (row.key === 'notification_config') merged.notifications = row.value;
          if (row.key === 'system_config') merged.system = row.value;
          if (row.key === 'admin_profile') merged.admin_profile = row.value;
        }
        return transformFromBackend(merged);
      }
    } catch (sErr) {
      console.warn('[settingsService] getAdminSettings Supabase direct fallback:', sErr);
    }
  }

  return DEFAULT_FULL_SETTINGS;
}

export async function updateAdminSettings(updates: Partial<FullSettings>): Promise<FullSettings> {
  const payload = transformToBackend(updates);
  try {
    const res = await api.admin.updateSettings(payload);
    if (res && res.data) {
      return transformFromBackend(res.data);
    }
  } catch (err) {
    console.warn('[settingsService] updateAdminSettings FastAPI failed, attempting Supabase direct:', err);
    if (isSupabaseConfigured && supabase) {
      if (payload.event) {
        await supabase.from('system_settings').upsert({
          key: 'event_config',
          value: payload.event,
          updated_by: 'admin',
        }, { onConflict: 'key' });
      }
      if (payload.notifications) {
        await supabase.from('system_settings').upsert({
          key: 'notification_config',
          value: payload.notifications,
          updated_by: 'admin',
        }, { onConflict: 'key' });
      }
      if (payload.system) {
        await supabase.from('system_settings').upsert({
          key: 'system_config',
          value: payload.system,
          updated_by: 'admin',
        }, { onConflict: 'key' });
      }
      return getAdminSettings();
    }
    throw err;
  }
  return getAdminSettings();
}

export async function resetAdminSettings(section?: string): Promise<FullSettings> {
  try {
    const res = await api.admin.resetSettings(section);
    if (res && res.data) {
      return transformFromBackend(res.data);
    }
  } catch (err) {
    console.warn('[settingsService] resetAdminSettings FastAPI failed:', err);
  }
  return DEFAULT_FULL_SETTINGS;
}

// ─── USER ROLES ─────────────────────────────────────────────────────────────

export async function getUserRoles(): Promise<UserRoleItem[]> {
  try {
    const res = await api.admin.getRoles();
    if (res && res.data && Array.isArray(res.data)) {
      return res.data.map((r: any) => ({
        id: String(r.id),
        userEmail: r.user_email,
        role: r.role,
        displayName: r.display_name || '',
        department: r.department || '',
        isActive: r.is_active ?? true,
        assignedBy: r.assigned_by || 'admin',
        createdAt: r.created_at,
        updatedAt: r.updated_at,
      }));
    }
  } catch (err) {
    console.warn('[settingsService] getUserRoles FastAPI fallback:', err);
  }

  if (isSupabaseConfigured && supabase) {
    try {
      const { data, error } = await supabase.from('user_roles').select('*');
      if (!error && data && data.length > 0) {
        return data.map((r: any) => ({
          id: String(r.id),
          userEmail: r.user_email,
          role: r.role,
          displayName: r.display_name || '',
          department: r.department || '',
          isActive: r.is_active ?? true,
          assignedBy: r.assigned_by || 'admin',
          createdAt: r.created_at,
          updatedAt: r.updated_at,
        }));
      }
    } catch (sErr) {
      console.warn('[settingsService] getUserRoles Supabase direct fallback:', sErr);
    }
  }

  return DEFAULT_USER_ROLES;
}

export async function createUserRole(data: {
  userEmail: string;
  role: 'admin' | 'coordinator' | 'jury' | 'participant';
  displayName?: string;
  department?: string;
  isActive?: boolean;
}): Promise<UserRoleItem> {
  const payload = {
    user_email: data.userEmail.trim().toLowerCase(),
    role: data.role,
    display_name: data.displayName || '',
    department: data.department || '',
    is_active: data.isActive ?? true,
  };

  try {
    const res = await api.admin.createRole(payload);
    if (res && res.data) {
      return {
        id: String(res.data.id),
        userEmail: res.data.user_email,
        role: res.data.role,
        displayName: res.data.display_name,
        department: res.data.department,
        isActive: res.data.is_active,
        assignedBy: res.data.assigned_by,
      };
    }
  } catch (err) {
    console.warn('[settingsService] createUserRole FastAPI failed, attempting Supabase direct:', err);
    if (isSupabaseConfigured && supabase) {
      const { data: supaData, error } = await supabase.from('user_roles').insert([payload]).select().single();
      if (!error && supaData) {
        return {
          id: String(supaData.id),
          userEmail: supaData.user_email,
          role: supaData.role,
          displayName: supaData.display_name,
          department: supaData.department,
          isActive: supaData.is_active,
          assignedBy: supaData.assigned_by,
        };
      }
      if (error) throw new Error(`Database error: ${error.message}`);
    }
    throw err;
  }

  return {
    id: `role-${Date.now()}`,
    userEmail: data.userEmail,
    role: data.role,
    displayName: data.displayName || '',
    department: data.department || '',
    isActive: data.isActive ?? true,
    assignedBy: 'admin',
  };
}

export async function updateUserRole(id: string, updates: Partial<UserRoleItem>): Promise<void> {
  const payload: Record<string, any> = {};
  if (updates.role !== undefined) payload.role = updates.role;
  if (updates.displayName !== undefined) payload.display_name = updates.displayName;
  if (updates.department !== undefined) payload.department = updates.department;
  if (updates.isActive !== undefined) payload.is_active = updates.isActive;

  try {
    await api.admin.updateRole(id, payload);
    return;
  } catch (err) {
    console.warn('[settingsService] updateUserRole FastAPI failed, attempting Supabase direct:', err);
    if (isSupabaseConfigured && supabase) {
      const { error } = await supabase.from('user_roles').update(payload).eq('id', id);
      if (!error) return;
      throw new Error(`Database error: ${error.message}`);
    }
    throw err;
  }
}

export async function deleteUserRole(id: string): Promise<void> {
  try {
    await api.admin.deleteRole(id);
    return;
  } catch (err) {
    console.warn('[settingsService] deleteUserRole FastAPI failed, attempting Supabase direct:', err);
    if (isSupabaseConfigured && supabase) {
      const { error } = await supabase.from('user_roles').delete().eq('id', id);
      if (!error) return;
      throw new Error(`Database error: ${error.message}`);
    }
    throw err;
  }
}

// ─── AUDIT LOGS ─────────────────────────────────────────────────────────────

export async function getAuditLogs(limit: number = 30): Promise<AuditLogItem[]> {
  try {
    const res = await api.admin.getAuditLogs(limit);
    if (res && res.data && Array.isArray(res.data)) {
      return res.data.map((l: any) => ({
        id: String(l.id),
        action: l.action,
        performedBy: l.performed_by,
        target: l.target,
        details: l.details,
        createdAt: l.created_at,
      }));
    }
  } catch (err) {
    console.warn('[settingsService] getAuditLogs FastAPI fallback:', err);
  }

  if (isSupabaseConfigured && supabase) {
    try {
      const { data, error } = await supabase
        .from('audit_logs')
        .select('*')
        .order('created_at', { ascending: false })
        .limit(limit);
      if (!error && data) {
        return data.map((l: any) => ({
          id: String(l.id),
          action: l.action,
          performedBy: l.performed_by,
          target: l.target,
          details: l.details,
          createdAt: l.created_at,
        }));
      }
    } catch (sErr) {
      console.warn('[settingsService] getAuditLogs Supabase direct fallback:', sErr);
    }
  }

  return [];
}

File: backend/app/services/settings_service.py
import uuid
import datetime
from typing import Dict, Any, List, Optional
from app.database import db
from app.schemas.settings import (
    FullSettings,
    EventConfig,
    NotificationConfig,
    SystemConfig,
    AdminProfile,
    FullSettingsUpdate,
    UserRoleItem,
    UserRoleCreate,
    UserRoleUpdate,
    AuditLogItem
)

DEFAULT_EVENT_CONFIG = {
    "event_name": "PRAGATHI 2K26",
    "event_date": "09 October 2026",
    "target_date_iso": "2026-10-09T09:00:00+05:30",
    "registration_status": "open",
    "registration_open_date": "2026-08-01T00:00:00+05:30",
    "registration_close_date": "2026-10-01T23:59:59+05:30",
    "website_visibility": "published",
    "event_status": "active"
}

DEFAULT_NOTIFICATION_CONFIG = {
    "announcement_emails_enabled": True,
    "registration_confirmation_emails_enabled": True,
    "event_reminder_alerts_enabled": True,
    "email_sender_name": "PRAGATHI 2K26 Secretariat",
    "provider_status": "ready"
}

DEFAULT_SYSTEM_CONFIG = {
    "maintenance_mode": False,
    "max_registrations": 500,
    "announcement_banner_enabled": False,
    "announcement_banner_text": "Welcome to PRAGATHI 2K26 Expo Registration Portal!",
    "debug_logging": False
}

DEFAULT_ADMIN_PROFILE = {
    "display_name": "Lead Administrator",
    "email": "admin@sru.edu.in",
    "role": "admin",
    "account_status": "verified",
    "last_sign_in": None,
    "auth_provider": "email"
}

DEFAULT_USER_ROLES = [
    {
        "id": "role-admin-1",
        "user_email": "admin@sru.edu.in",
        "role": "admin",
        "display_name": "Lead Administrator",
        "department": "Deanery",
        "is_active": True,
        "assigned_by": "system",
        "created_at": "2026-08-01T00:00:00Z"
    },
    {
        "id": "role-coord-1",
        "user_email": "coordinator@sru.edu.in",
        "role": "coordinator",
        "display_name": "Chief Event Coordinator",
        "department": "CSE",
        "is_active": True,
        "assigned_by": "system",
        "created_at": "2026-08-01T00:00:00Z"
    },
    {
        "id": "role-jury-1",
        "user_email": "jury.head@sru.edu.in",
        "role": "jury",
        "display_name": "Evaluation Panel Head",
        "department": "Research",
        "is_active": True,
        "assigned_by": "system",
        "created_at": "2026-08-01T00:00:00Z"
    }
]

class SettingsService:
    def __init__(self):
        # Ensure local DB has system_settings, user_roles, audit_logs initialized
        local_data = db.load_local()
        updated = False
        if "system_settings" not in local_data:
            local_data["system_settings"] = {
                "event_config": DEFAULT_EVENT_CONFIG.copy(),
                "notification_config": DEFAULT_NOTIFICATION_CONFIG.copy(),
                "system_config": DEFAULT_SYSTEM_CONFIG.copy(),
                "admin_profile": DEFAULT_ADMIN_PROFILE.copy()
            }
            updated = True
        if "user_roles" not in local_data:
            local_data["user_roles"] = DEFAULT_USER_ROLES.copy()
            updated = True
        if "audit_logs" not in local_data:
            local_data["audit_logs"] = [
                {
                    "id": "audit-init",
                    "action": "SYSTEM_INIT",
                    "performed_by": "system",
                    "target": "system_settings",
                    "details": "System settings and security framework initialized.",
                    "created_at": datetime.datetime.now(datetime.timezone.utc).isoformat()
                }
            ]
            updated = True
        if updated:
            db.save_local(local_data)

    async def get_settings(self) -> FullSettings:
        # Try fetching from Supabase system_settings table
        supa_settings = await db.fetch_supabase("system_settings")
        event_dict = DEFAULT_EVENT_CONFIG.copy()
        notif_dict = DEFAULT_NOTIFICATION_CONFIG.copy()
        sys_dict = DEFAULT_SYSTEM_CONFIG.copy()
        admin_dict = DEFAULT_ADMIN_PROFILE.copy()

        if supa_settings and isinstance(supa_settings, list):
            for row in supa_settings:
                key = row.get("key")
                val = row.get("value") or {}
                if key == "event_config":
                    event_dict.update(val)
                elif key == "notification_config":
                    notif_dict.update(val)
                elif key == "system_config":
                    sys_dict.update(val)
                elif key == "admin_profile":
                    admin_dict.update(val)
        else:
            local = db.load_local().get("system_settings", {})
            event_dict.update(local.get("event_config", {}))
            notif_dict.update(local.get("notification_config", {}))
            sys_dict.update(local.get("system_config", {}))
            admin_dict.update(local.get("admin_profile", {}))

        return FullSettings(
            event=EventConfig(**event_dict),
            notifications=NotificationConfig(**notif_dict),
            system=SystemConfig(**sys_dict),
            admin_profile=AdminProfile(**admin_dict)
        )

    async def update_settings(self, updates: FullSettingsUpdate, updated_by: str = "admin") -> FullSettings:
        current = await self.get_settings()
        local_data = db.load_local()
        local_settings = local_data.get("system_settings", {})

        changed_sections = []

        if updates.event is not None:
            new_event = updates.event.model_dump()
            local_settings["event_config"] = new_event
            await db.upsert_supabase("system_settings", {
                "key": "event_config",
                "value": new_event,
                "description": "Core event lifecycle and registration window",
                "is_public": True,
                "updated_by": updated_by
            }, on_conflict="key")
            changed_sections.append("event_config")

        if updates.notifications is not None:
            new_notif = updates.notifications.model_dump()
            local_settings["notification_config"] = new_notif
            await db.upsert_supabase("system_settings", {
                "key": "notification_config",
                "value": new_notif,
                "description": "Email delivery and notification flags",
                "is_public": False,
                "updated_by": updated_by
            }, on_conflict="key")
            changed_sections.append("notification_config")

        if updates.system is not None:
            new_sys = updates.system.model_dump()
            local_settings["system_config"] = new_sys
            await db.upsert_supabase("system_settings", {
                "key": "system_config",
                "value": new_sys,
                "description": "System-level limits, banner, and maintenance mode",
                "is_public": True,
                "updated_by": updated_by
            }, on_conflict="key")
            changed_sections.append("system_config")

        if updates.admin_profile is not None:
            admin_dict = current.admin_profile.model_dump()
            admin_dict.update(updates.admin_profile)
            local_settings["admin_profile"] = admin_dict
            await db.upsert_supabase("system_settings", {
                "key": "admin_profile",
                "value": admin_dict,
                "description": "Admin profile metadata",
                "is_public": False,
                "updated_by": updated_by
            }, on_conflict="key")
            changed_sections.append("admin_profile")

        local_data["system_settings"] = local_settings
        db.save_local(local_data)

        # Log audit action
        await self.log_audit_action(
            action="SETTINGS_UPDATE",
            performed_by=updated_by,
            target="system_settings",
            details=f"Updated configuration sections: {', '.join(changed_sections)}"
        )

        return await self.get_settings()

    async def reset_settings(self, section: Optional[str] = None, updated_by: str = "admin") -> FullSettings:
        local_data = db.load_local()
        local_settings = local_data.get("system_settings", {})

        if not section or section == "all":
            local_settings["event_config"] = DEFAULT_EVENT_CONFIG.copy()
            local_settings["notification_config"] = DEFAULT_NOTIFICATION_CONFIG.copy()
            local_settings["system_config"] = DEFAULT_SYSTEM_CONFIG.copy()
            local_settings["admin_profile"] = DEFAULT_ADMIN_PROFILE.copy()
            detail = "All system settings restored to default baseline."
        elif section == "event":
            local_settings["event_config"] = DEFAULT_EVENT_CONFIG.copy()
            detail = "Event configuration restored to default baseline."
        elif section == "notifications":
            local_settings["notification_config"] = DEFAULT_NOTIFICATION_CONFIG.copy()
            detail = "Notification configuration restored to default baseline."
        elif section == "system":
            local_settings["system_config"] = DEFAULT_SYSTEM_CONFIG.copy()
            detail = "System configuration restored to default baseline."

        local_data["system_settings"] = local_settings
        db.save_local(local_data)

        # Sync with Supabase
        for key in ["event_config", "notification_config", "system_config", "admin_profile"]:
            val = local_settings.get(key)
            if val:
                await db.upsert_supabase("system_settings", {
                    "key": key,
                    "value": val,
                    "updated_by": updated_by
                }, on_conflict="key")

        await self.log_audit_action(
            action="SETTINGS_RESET",
            performed_by=updated_by,
            target=section or "all",
            details=detail
        )

        return await self.get_settings()

    # ─── USER ROLES CRUD ───────────────────────────────────────────────────────

    async def get_roles(self) -> List[UserRoleItem]:
        supa_roles = await db.fetch_supabase("user_roles")
        if supa_roles and isinstance(supa_roles, list) and len(supa_roles) > 0:
            return [
                UserRoleItem(
                    id=str(r.get("id")),
                    user_email=r.get("user_email", ""),
                    role=r.get("role", "participant"),
                    display_name=r.get("display_name", ""),
                    department=r.get("department", ""),
                    is_active=r.get("is_active", True),
                    assigned_by=r.get("assigned_by", "admin"),
                    created_at=r.get("created_at"),
                    updated_at=r.get("updated_at")
                )
                for r in supa_roles
            ]

        local_roles = db.load_local().get("user_roles", DEFAULT_USER_ROLES)
        return [UserRoleItem(**r) for r in local_roles]

    async def create_role(self, data: UserRoleCreate, assigned_by: str = "admin") -> UserRoleItem:
        role_id = f"role-{uuid.uuid4().hex[:8]}"
        now = datetime.datetime.now(datetime.timezone.utc).isoformat()
        new_item = {
            "id": role_id,
            "user_email": data.user_email.strip().lower(),
            "role": data.role,
            "display_name": data.display_name.strip() if data.display_name else "",
            "department": data.department.strip() if data.department else "",
            "is_active": data.is_active,
            "assigned_by": assigned_by,
            "created_at": now,
            "updated_at": now
        }

        # Try Supabase
        supa_res = await db.insert_supabase("user_roles", new_item)
        if supa_res and isinstance(supa_res, dict):
            new_item["id"] = str(supa_res.get("id", role_id))

        # Local storage sync
        local_data = db.load_local()
        roles = local_data.setdefault("user_roles", [])
        # Check if email already exists locally
        roles = [r for r in roles if r.get("user_email") != new_item["user_email"]]
        roles.append(new_item)
        local_data["user_roles"] = roles
        db.save_local(local_data)

        await self.log_audit_action(
            action="ROLE_ASSIGNED",
            performed_by=assigned_by,
            target=new_item["user_email"],
            details=f"Assigned role '{new_item['role']}' to {new_item['user_email']}"
        )

        return UserRoleItem(**new_item)

    async def update_role(self, role_id: str, data: UserRoleUpdate, performed_by: str = "admin") -> Optional[UserRoleItem]:
        local_data = db.load_local()
        roles = local_data.setdefault("user_roles", [])
        target = next((r for r in roles if str(r.get("id")) == str(role_id)), None)

        if not target:
            # Check supabase
            supa_roles = await db.fetch_supabase("user_roles")
            if supa_roles and isinstance(supa_roles, list):
                target = next((r for r in supa_roles if str(r.get("id")) == str(role_id)), None)

        if not target:
            return None

        if data.role is not None:
            target["role"] = data.role
        if data.display_name is not None:
            target["display_name"] = data.display_name.strip()
        if data.department is not None:
            target["department"] = data.department.strip()
        if data.is_active is not None:
            target["is_active"] = data.is_active

        target["updated_at"] = datetime.datetime.now(datetime.timezone.utc).isoformat()

        # Update in Supabase
        await db.upsert_supabase("user_roles", target, on_conflict="id")

        # Update in Local JSON
        for i, r in enumerate(roles):
            if str(r.get("id")) == str(role_id):
                roles[i] = target
                break
        db.save_local(local_data)

        await self.log_audit_action(
            action="ROLE_UPDATED",
            performed_by=performed_by,
            target=target.get("user_email", role_id),
            details=f"Updated role configuration for {target.get('user_email')}"
        )

        return UserRoleItem(**target)

    async def delete_role(self, role_id: str, performed_by: str = "admin") -> bool:
        local_data = db.load_local()
        roles = local_data.setdefault("user_roles", [])
        target = next((r for r in roles if str(r.get("id")) == str(role_id)), None)

        user_email = target.get("user_email", role_id) if target else role_id

        # Delete from local
        local_data["user_roles"] = [r for r in roles if str(r.get("id")) != str(role_id)]
        db.save_local(local_data)

        # Delete from Supabase
        await db.delete_supabase("user_roles", "id", str(role_id))

        await self.log_audit_action(
            action="ROLE_REVOKED",
            performed_by=performed_by,
            target=user_email,
            details=f"Revoked role assignment for {user_email}"
        )

        return True

    # ─── AUDIT LOGS ───────────────────────────────────────────────────────────

    async def get_audit_logs(self, limit: int = 30) -> List[AuditLogItem]:
        supa_logs = await db.fetch_supabase("audit_logs")
        if supa_logs and isinstance(supa_logs, list) and len(supa_logs) > 0:
            sorted_logs = sorted(supa_logs, key=lambda x: x.get("created_at", ""), reverse=True)
            return [
                AuditLogItem(
                    id=str(log.get("id")),
                    action=log.get("action", "UNKNOWN"),
                    performed_by=log.get("performed_by", "admin"),
                    target=log.get("target", ""),
                    details=log.get("details", ""),
                    created_at=log.get("created_at")
                )
                for log in sorted_logs[:limit]
            ]

        local_logs = db.load_local().get("audit_logs", [])
        sorted_local = sorted(local_logs, key=lambda x: x.get("created_at", ""), reverse=True)
        return [AuditLogItem(**l) for l in sorted_local[:limit]]

    async def log_audit_action(self, action: str, performed_by: str, target: str = "", details: str = ""):
        log_entry = {
            "id": f"audit-{uuid.uuid4().hex[:8]}",
            "action": action,
            "performed_by": performed_by,
            "target": target,
            "details": details,
            "created_at": datetime.datetime.now(datetime.timezone.utc).isoformat()
        }

        # Try Supabase insert
        await db.insert_supabase("audit_logs", log_entry)

        # Local storage insert
        local_data = db.load_local()
        logs = local_data.setdefault("audit_logs", [])
        logs.insert(0, log_entry)
        if len(logs) > 100:
            local_data["audit_logs"] = logs[:100]
        db.save_local(local_data)

settings_service = SettingsService()
